posthog/special_migrations/migrations/example.py

```diff
@@ -1 +1,112 @@
+    #   1.1. receive the output of the previous op?
+    operations = [
+        SpecialMigrationOperation(
+            database="clickhouse",
+            sql=PERSONS_DISTINCT_ID_TABLE_SQL.replace(PERSONS_DISTINCT_ID_TABLE, TEMPORARY_TABLE_NAME, 1),
+            rollback=f"DROP TABLE {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+        ),
+        SpecialMigrationOperation(
+            database="clickhouse",
+            sql=f"DROP TABLE person_distinct_id_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=PERSONS_DISTINCT_ID_TABLE_MV_SQL,
+        ),
+        SpecialMigrationOperation(
+            database="clickhouse",
+            sql=f"DROP TABLE kafka_person_distinct_id ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=KAFKA_PERSONS_DISTINCT_ID_TABLE_SQL,
+        ),
+        SpecialMigrationOperation(
+            database="clickhouse",
+            timeout_seconds=ONE_DAY,
+            sql=f"""
+                INSERT INTO {TEMPORARY_TABLE_NAME} (distinct_id, person_id, team_id, _sign, _timestamp, _offset)
+                SELECT
+                    distinct_id,
+                    person_id,
+                    team_id,
+                    if(is_deleted==0, 1, -1) as _sign,
+                    _timestamp,
+                    _offset
+                FROM {PERSONS_DISTINCT_ID_TABLE}
+            """,
+            rollback="",
+        ),
+        SpecialMigrationOperation(
+            database="clickhouse",
+            sql=f"""
+                RENAME TABLE
+                    {CLICKHOUSE_DATABASE}.{PERSONS_DISTINCT_ID_TABLE} to {CLICKHOUSE_DATABASE}.person_distinct_id_backup,
+                    {CLICKHOUSE_DATABASE}.{TEMPORARY_TABLE_NAME} to {CLICKHOUSE_DATABASE}.{PERSONS_DISTINCT_ID_TABLE}
+                ON CLUSTER {CLICKHOUSE_CLUSTER}
+            """,
+        ),
+        SpecialMigrationOperation(database="clickhouse", sql=KAFKA_PERSONS_DISTINCT_ID_TABLE_SQL,),
+        SpecialMigrationOperation(database="clickhouse", sql=PERSONS_DISTINCT_ID_TABLE_MV_SQL,),
+    ]
+
+    def healthcheck(self):
+        result = sync_execute("SELECT total_space, free_space FROM system.disks")
+        total_space = result[0][0]
+        free_space = result[0][1]
+        if free_space > total_space / 2:
+            return (True, None)
+        else:
+            return (False, "Upgrade your ClickHouse storage.")
+
+    def progress(self, _):
+        result = sync_execute(f"SELECT COUNT(1) FROM {TEMPORARY_TABLE_NAME}")
+        result2 = sync_execute(f"SELECT COUNT(1) FROM {PERSONS_DISTINCT_ID_TABLE}")
+        total_events_to_move = result2[0][0]
+        total_events_moved = result[0][0]
+
+        progress = 100 * total_events_moved / total_events_to_move
+        return progress
+
+    def rollback(self, migration_instance):
+        current_operation_index = migration_instance.current_operation_index
+        if current_operation_index > 4:
+            sync_execute(
```

Comment:
I think this section would be cleaner and the overall code safer of this method didn't contain any logic regarding how to perform the rollback. 

In my opinion this should be a blind for loop issuing the rollback statements described above

Reply:
Some discussion happened here regarding this:

https://github.com/PostHog/posthog/pull/7442#discussion_r759524547

Reply:
This brings up the point about `current_operation_index` update timing : we update the index after we've completed the step (in the code in the original WIP), so the op could be partially done. Imagine a plugin server restart at a suboptimal time. Furthermore this could be a problem for the migration itself too, e.g. if we copy data twice we might not have enough space left or worse have dupe data we shouldn't have (your insert into step here). So it might be safest if we do the op & current_operation_index update in a transaction both ways.

Reply:
Do you mean running the operation in a transaction? If so, that's what is already done in Postgres, but CH doesn't have transactions

Reply:
updating the op index & running the sql command in the step either both or neither (rollback for both)